app/models/miq_region.rb

```diff
@@ -242,6 +243,59 @@ def remote_ws_url
     hostname.nil? ? nil : "https://#{hostname}"
   end
 
+  def generate_auth_key_queue(ssh_user, ssh_password, ssh_host = nil)
+    args = [ssh_user, MiqPassword.try_encrypt(ssh_password)]
+    args << ssh_host if ssh_host
+
+    MiqQueue.put_unless_exists(
+      :class_name  => self.class.name,
+      :instance_id => id,
+      :queue_name  => "generic",
+      :method_name => "generate_auth_key",
+      :args        => args
+    )
+  end
+
+  def generate_auth_key(ssh_user, ssh_password, ssh_host = remote_ws_address)
+    key = remote_region_v2_key(ssh_user, MiqPassword.try_decrypt(ssh_password), ssh_host)
+
+    auth = AuthToken.new
+    auth.auth_key = key
+    auth.name = "Region #{region} API Key"
+    auth.resource = self
+    auth.authtype = "system_api"
+    auth.save!
+  end
+
+  def verify_credentials(_auth_type = nil, _options = nil)
+    # TODO: verify the key against the remote api using the api client gem
+    true
+  end
+
+  def api_system_auth_token(userid)
+    region_v2_key = authentication_token("system_api")
+
+    token_hash = {
+      :server_guid => remote_ws_miq_server.guid,
+      :userid      => userid,
+      :timestamp   => Time.now.utc
+    }
+
+    file = Tempfile.new("region_auth_key")
+    begin
+      file.write(region_v2_key)
+      file.close
+      key = EzCrypto::Key.load(file.path)
+      MiqPassword.new.encrypt(token_hash.to_yaml, "v2", key)
+    ensure
+      file.unlink
+    end
+  end
+
+  def self.api_system_auth_token_for_region(region_id)
```

Comment:
oops, this needs to be passed the userid to use for the token and the method call below will just blow up as is.